### ISSUE TITLE
Running tests - multiple stores

### ISSUE THREAD

**Issue by phoe** [hanshuebner/bknr-datastore#11]
Let's say that I have some persistent classes.
BKNR supports only a single data store at a given time.
While I have a server running with the production store loaded, I want to be able to run unit tests without test data being written into a production store; I would like them to end up in some test store that I can discard once the tests are done.
Can I use the `WITH-STORE` macro found in https://github.com/hanshuebner/bknr-datastore/blob/master/src/data/txn.lisp#L97 for this?

**Comment by hanshuebner** [comment]
I think you can use `WITH-STORE` for the purpose you describe.  I must admit, though, that I have not used the macro in a long time and can't really give good and complete advice.  If in doubt, you need to read the source.

**Comment by phoe** [comment]
Okay - I'll experiment with it during the weekend and let you know how that goes.

From what I see, everything in BKNR.DATASTORE accesses (or *should* access) the store through the `*STORE*` dynamic variable - and if that's the case, then I can shadow it with another value and therefore another store.

**Comment by phoe** [comment]
Before I move on with testing on my own, I have a question about organizing multiple stores.

I have an example implementation of a game server. The data I want to contain in there is formed like a tree:
```
* worlds
  + terrain
  + players
  * chats
    + messages
```

So basically, I want to have multiple worlds. I want to contain two example worlds under pathnames `/world/foo-world/` and `/world/quux-world/`.

Inside a world, I can have data about terrain and people who live on it. So, I want to have two stores for them - they will have pathnames `/world/foo-world/terrain/` and `/world/foo-world/players/`. 

A world can also have chats, which are histories of messages. I want two example chats to have stores under pathnames `/worlds/foo-world/chats/bar-chat/` and `/worlds/foo-world/chats/baz-chat/`.

Summing up - I want my data to reside under these pathnames:
```
/worlds/foo-world/terrain/*
/worlds/foo-world/players/*
/worlds/foo-world/chats/bar-chat/*
/worlds/foo-world/chats/baz-chat/*
/worlds/quux-world/*
```

Can the current capabilities of BKNR.DATASTORE help me achieve such a goal?

**Comment by hanshuebner** [comment]
It was not a design goal for the datastore to support multiple disparate
worlds within one running Lisp image.  It is not unlikely that `with-store`
and the `*store*` dynamic variable will help you in having separate worlds,
but be prepared that you may find leaks that you have to close yourself.

2017-01-24 12:06 GMT-05:00 Michał Herda <notifications@github.com>:

**Comment by phoe** [comment]
Thanks. I've begun biting into it and I'll let you know how it goes.

**Comment by tdrhq** [comment]
@phoe I'm about to try to use with-store for tests for the same purpose as you did. Any findings I should be warned about?

**Comment by phoe** [comment]
None that I remember anymore.

**Comment by tdrhq** [comment]
@phoe Thanks!
